File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        result = []

        def combo(noder, output):
            output.append(noder.val)
            if noder.left:
                combo(noder.left, output)
            if noder.right:
                combo(noder.right, output)
            
        if root:
            combo(root, result)
        return result

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> answer = new ArrayList<>();
        if(root != null){
            traverse(root, answer);
        }
        return answer;
        
    }
    
    private void traverse(TreeNode root, List<Integer> lister){
        
        lister.add(root.val);
        
        if(root.left != null){
            traverse(root.left, lister);
        }
        if(root.right != null){
            traverse(root.right, lister);
        }
        
    }
    
}